e2e/e2e_partition_test.go

```diff
@@ -1,14 +1,133 @@
+// This test creates a single cluster of 5 nodes, but instead of letting all the peers communicate with each other,
+// it routes messages only to specific nodes and induces that nodes lock on different proposal.
+// In the round=1, it marks one node as faulty (meaning that it doesn't takes part in gossiping).
+func TestE2E_Partition_Liveness(t *testing.T) {
+	const nodesCnt = 5
+	round0 := roundMetadata{
+		round: 0,
+		// induce locking A_3 and A_4 on one proposal
+		routingMap: map[sender]receivers{
+			"A_0": {"A_3", "A_4"},
+			"A_3": {"A_0", "A_3", "A_4"},
+			"A_4": {"A_3", "A_4"},
+		},
+	}
+	round1 := roundMetadata{
+		round: 1,
+		// induce locking lock A_0 and A_2 on another proposal
+		routingMap: map[sender]receivers{
+			"A_0": {"A_0", "A_2", "A_3", "A_4"},
+			"A_1": {"A_0", "A_2", "A_3", "A_4"},
+			"A_2": {"A_0", "A_1", "A_2", "A_3", "A_4"},
+
+			"A_3": {"A_0", "A_1", "A_2", "A_3", "A_4"},
+			"A_4": {"A_0", "A_1", "A_2", "A_3", "A_4"},
+		},
+	}
+	flowMap := map[uint64]roundMetadata{0: round0, 1: round1}
+
+	faultyNodeId := pbft.NodeID("A_1")
+
+	// If livenessGossipHandler returns false, message should not be transported.
+	livenessGossipHandler := func(senderId, receiverId pbft.NodeID, msg *pbft.MessageReq) bool {
+		if msg.View.Round > 1 || msg.View.Sequence > 2 {
+			// Faulty node is unresponsive after round 1, and all the other nodes are gossiping all the messages.
+			return senderId != faultyNodeId && receiverId != faultyNodeId
+		} else {
+			if msg.View.Round <= 1 && msg.Type == pbft.MessageReq_Commit {
+				// Cut all the commit messages gossiping for round 0 and 1
+				return false
+			}
+			if msg.View.Round == 1 && senderId == faultyNodeId &&
+				(msg.Type == pbft.MessageReq_RoundChange || msg.Type == pbft.MessageReq_Commit) {
+				// Case where we are in round 1 and 2 different nodes will lock the proposal
+				// (consequence of faulty node doesn't gossip round change and commit messages).
+				return false
+			}
+		}
+
+		return shouldGossipBasedOnMsgFlowMap(flowMap, msg, senderId, receiverId)
+	}
+
+	hook := newGenericGossipTransport(livenessGossipHandler)
+
+	c := newPBFTCluster(t, "liveness_issue", "A", nodesCnt, hook)
+	c.Start()
+	defer c.Stop()
+
+	err := c.WaitForHeight(3, 5*time.Minute)
+
+	// Query nodes and make appropriate assertions
+	nodeSubsetA := []pbft.NodeID{"A_0", "A_1", "A_2"}
+	nodeSubsetB := []pbft.NodeID{"A_3", "A_4"}
+	nodeA0 := c.nodes[string(nodeSubsetA[0])]
+	nodeA3 := c.nodes[string(nodeSubsetB[0])]
+	queryNodesFromSubset(t, c.nodes, faultyNodeId, nodeA0, nodeSubsetA)
+	queryNodesFromSubset(t, c.nodes, faultyNodeId, nodeA3, nodeSubsetB)
+
+	assert.NoError(t, err)
+}
+
+// Helper function determining whether a message should be gossiped, based on round message flow.
+func shouldGossipBasedOnMsgFlowMap(flowMap map[uint64]roundMetadata, msg *pbft.MessageReq, senderId pbft.NodeID, receiverId pbft.NodeID) bool {
```

Comment:
Could you add this function as a method of `roundMetadata`?
```
func (r *roundMetadata) shouldGossipBasedOnMsgFlowMap...
```

Reply:
Seems like it more suits to the `genericGossipTransport`, so I have moved it there. Hope that you don't mind.